### PR TITLE
Check if building's dependencies are allowed

### PR DESCRIPTION
**Pull request by Chocimier**
Fix for bug 2321

### CODE REVIEW

File: lib/CGameInfoCallback.cpp [hunk 1]
```diff
@@ -520,6 +520,28 @@ EBuildingState::EBuildingState CGameInfoCallback::canBuildStructure( const CGTow
 	if(vstd::contains(t->forbiddenBuildings, ID))
 		return EBuildingState::FORBIDDEN; //forbidden
 
+	std::function<bool(BuildingID id)> allowedTest;
+	std::function<bool(BuildingID id)> possiblyNotBuiltTest;
+
+	allowedTest = [&](BuildingID id) -> bool
```
Comment (alexvins) [inline lib/CGameInfoCallback.cpp]: We declare lambdas as "auto"

File: lib/CGameInfoCallback.cpp [hunk 2]
```diff
@@ -520,6 +520,28 @@ EBuildingState::EBuildingState CGameInfoCallback::canBuildStructure( const CGTow
...
+
+		return t->genBuildingRequirements(id, true).satisfiable(allowedTest, possiblyNotBuiltTest);
+	};
+
+	possiblyNotBuiltTest = [&](BuildingID id) -> bool
+	{
+		//TODO consider destroing
+		return !t->hasBuilt(id);
+	};
+
```
Comment (alexvins) [inline lib/CGameInfoCallback.cpp]: Define this before allowedTest

File: lib/CGameInfoCallback.cpp [hunk 3]
```diff
@@ -520,6 +520,28 @@ EBuildingState::EBuildingState CGameInfoCallback::canBuildStructure( const CGTow
...
+		{
+			return false;
+		}
+
+		return t->genBuildingRequirements(id, true).satisfiable(allowedTest, possiblyNotBuiltTest);
+	};
+
+	possiblyNotBuiltTest = [&](BuildingID id) -> bool
+	{
+		//TODO consider destroing
```
Comment (alexvins) [inline lib/CGameInfoCallback.cpp]: Please explain this TODO item

File: lib/LogicalExpression.h [hunk 4]
```diff
@@ -99,6 +99,123 @@ namespace LogicalExpressionDetail
...
+	class FalsifiabilityVisitor;
+
+	/// Visitor to test whether expression's value can be true
+	template <typename ContainedClass>
+	class SatisfiabilityVisitor : public boost::static_visitor<bool>
+	{
+		typedef ExpressionBase<ContainedClass> Base;
+
+		std::function<bool(const typename Base::Value &)> satisfiabilityTest;
+		FalsifiabilityVisitor<ContainedClass> *falsifiabilityVisitor;
```
Comment (alexvins) [inline lib/LogicalExpression.h]: falsifiability[Visitor, Test], satisfiability[Visitor, Test], countSatisfiable, countFalsifiable should be extracted to common base class template.

smth like:

```
template <typename ContainedClass>
	class BaseVisitor : public boost::static_visitor<bool> ...

template <typename ContainedClass>
	class SatisfiabilityVisitor : public BaseVisitorContainedClass>
```

File: lib/CGameInfoCallback.cpp (hunk 3 above)
Comment (Chocimier) [inline lib/CGameInfoCallback.cpp]: This is about building A which need building B not to be built; can be expressed by genBuildingRequirements(A) = NoneOf(B).

If VCMI allow to destroy building, possiblyNotBuiltTest should rather return (!t->hasBuilt(id) || t->canBeDestroyed(id)) or just (true) if any building can be destroyed.

Assuming that is ever planned.

Comment (DjWarmonger) [inline lib/CGameInfoCallback.cpp]: - There are WoG scripts which allow to destroy building. Not a priority, but...
- In original game, Capitol is destroyed if player conquering a town already has one.

**Pull request merged**